### PR TITLE
feat: implement observability of notifications subsystem

### LINKED ISSUE #3: Implement observability of notifications subsystem
Relates to https://github.com/coder/coder/issues/13331

- [x] Notifications sent (by receiver, notification_template)
- [x] Notifications failed (by receiver, notification_template)
- [x] Notifications retried (by receiver, notification_template)
- [x] Notifications paused
- [ ] Notifications purged
- [ ] Queue operations
  - [ ] Enqueue time (delta between start of operation and accepted into db)
  - [x] Dequeue time (delta between enqueued and dequeued)
- [x] Receiver operations
  - [x] Send time (response time from server, by response status)

### PR DESCRIPTION
**Pull request by dannykopping**
Closes https://github.com/coder/team-coconut/issues/3

Also performed some drive-by refactorings.

_NOTE: this is stacked on `dk/system-notifications-lib, but I didn't want to use Graphite for fear it screws up that branch which is close to being merged; I'm still pretty new to that tool._

### CODE REVIEW

File: coderd/notifications/notifier.go
Comment (dannykopping) [inline coderd/notifications/notifier.go:213]: Reviewer: we may want to distinguish successful vs failed dispatch speed; what do you think?
I left it out because it's probably not very useful in practice.

File: coderd/database/migrations/000223_notifications_metrics.up.sql
Comment (mafredri) [inline coderd/database/migrations/000223_notifications_metrics.up.sql:2]: Nit: The name `queued_seconds` implies an integer, consider changing the type or the name.

File: coderd/notifications/manager.go
Comment (mafredri) [inline coderd/notifications/manager.go:63]: Will this be fixed in this PR, a follow-up, or should we track this via issue?

File: coderd/database/dbmem/dbmem.go
Comment (mtojek) [inline coderd/database/dbmem/dbmem.go:933]: why `sz` :) ?

File: coderd/notifications/manager.go
Comment (mafredri) [inline coderd/notifications/manager.go:66]: Nit: simplify (given is given 😄?)
```suggestion
		return nil, xerrors.Errorf("notification method %q is invalid", cfg.Method)
```

File: coderd/database/migrations/000223_notifications_metrics.up.sql
Comment (mtojek) [inline coderd/database/migrations/000223_notifications_metrics.up.sql:2]: Shouldn't we go with `numeric`?

File: coderd/notifications/metrics_test.go
Comment (mafredri) [inline coderd/notifications/metrics_test.go:160]: Note: Polling designs are often flakey, it would be nice if we could refactor it. Something to keep in mind for the future.

File: coderd/notifications/notifier.go
Comment (mafredri) [inline coderd/notifications/notifier.go:205]: I take it this is incremented somewhere not visible in this PR?

File: coderd/notifications/notifier.go
Comment (mtojek) [reply coderd/notifications/notifier.go:213]: IMHO Dispatching time is fine to be left as is, as a single timer. I would rather consider including the failure reason if possible.

**Review by mafredri — APPROVED** [review]
FWIW the changes LGTM in isolation (I haven't reviewed the original code).

File: coderd/notifications/metrics_test.go
Comment (mtojek) [inline coderd/notifications/metrics_test.go:31]: Is it because of pubsub or something else?

File: coderd/notifications/metrics.go
Comment (mtojek) [inline coderd/notifications/metrics.go:59]: DispatchingTimeSeconds :) ?

**Review by mtojek — APPROVED** [review]
Nothing blocking 👍

File: coderd/database/dbmem/dbmem.go
Comment (dannykopping) [reply coderd/database/dbmem/dbmem.go:933]: `size`

File: coderd/database/migrations/000223_notifications_metrics.up.sql
Comment (dannykopping) [reply coderd/database/migrations/000223_notifications_metrics.up.sql:2]: I don't think it necessarily implies integer; `time.Duration`'s [`Seconds()`](https://pkg.go.dev/time#Duration.Seconds) returns a float64.

File: coderd/notifications/manager.go
Comment (dannykopping) [reply coderd/notifications/manager.go:63]: In a follow-up, but the work is not yet planned.
This will need to be implemented once we expand beyond system-only notifications.
I think it's ok to leave untracked for now.

File: coderd/notifications/metrics.go
Comment (dannykopping) [reply coderd/notifications/metrics.go:59]: Seems a bit redundant to have "time" and "seconds" in the name.
I think this is already clear enough, TBH.

File: coderd/notifications/metrics_test.go
Comment (dannykopping) [reply coderd/notifications/metrics_test.go:31]: A lot of the business logic is implemented in the database layer.
I'm requiring postgres here because the full functionality is required (retries, etc).
I'll clarify this inline.

File: coderd/notifications/metrics_test.go
Comment (dannykopping) [reply coderd/notifications/metrics_test.go:160]: Indeed, I have https://github.com/coder/team-coconut/issues/6 scheduled.

File: coderd/notifications/notifier.go
Comment (dannykopping) [reply coderd/notifications/notifier.go:205]: Correct; in the database layer.

**Review by mtojek — APPROVED** [review]
Thanks for all replies.

File: coderd/database/dbmem/dbmem.go
Comment (dannykopping) [reply coderd/database/dbmem/dbmem.go:933]: I generally agree, although I try to follow this advice: https://go.dev/talks/2014/names.slide#4

> The greater the distance between a name's declaration and its uses, the longer the name should be.

Here it's pretty clear what `sz` does, so brevity is helpful.

Comment (mtojek) [reply coderd/database/dbmem/dbmem.go:933]: offtopic: It depends on your background. For me `sz` is reserved for **s**tring **z**ero-terminated (c++).

Comment (dannykopping) [reply coderd/database/dbmem/dbmem.go:933]: Ha! Fair enough :)

File: coderd/database/migrations/000223_notifications_metrics.up.sql
Comment (dannykopping) [reply coderd/database/migrations/000223_notifications_metrics.up.sql:2]: It's been a while since I've used PG so I had to look up the difference.

Seems like `FLOAT` might be more appropriate here since we don't need the absolute precision that `NUMERIC` provides and `FLOAT` is faster/smaller.

Is there some other reason I should consider using `NUMERIC` over `FLOAT` which I might be missing?

Comment (mtojek) [reply coderd/database/migrations/000223_notifications_metrics.up.sql:2]: Passing the database question to @mafredri. My understanding is that floats can always lose precision, so it is safer to store round numbers. If we need fractions of seconds, let's go with milliseconds?

Comment (mafredri) [reply coderd/database/migrations/000223_notifications_metrics.up.sql:2]: Fair enough 😄.

In the DB, usually it means seconds:

```
    lifetime_seconds bigint DEFAULT 86400 NOT NULL,
    timeout_seconds integer NOT NULL
    connection_timeout_seconds integer DEFAULT 0 NOT NULL,
```

But fine to keep as float if we care about the precision.

Comment (mafredri) [reply coderd/database/migrations/000223_notifications_metrics.up.sql:2]: I prefer ints as well, but not enough to block this PR, so I'll leave it up to Danny.

File: coderd/notifications/manager.go
Comment (mafredri) [reply coderd/notifications/manager.go:63]: The reason I lifted this up is that `// TODO`s in code is usually where TODO's go to die, which is why I prefer them either tracked or not important enough to mention. Not a blocker to leave it in, just a pet-peeve.

Exhibit A:

```
rg 'TODO\(mafredri\)' | wc -l
      31
```

File: coderd/database/migrations/000223_notifications_metrics.up.sql
Comment (dannykopping) [reply coderd/database/migrations/000223_notifications_metrics.up.sql:2]: I suppose I don't strictly _need_ to keep sub-second precision on the length of time messages were queued for.
I'll switch to `integer` 👍

Comment (dannykopping) [reply coderd/database/migrations/000223_notifications_metrics.up.sql:2]: As discussed in https://github.com/coder/coder/pull/13799#discussion_r1668521931, I actually don't specifically need sub-second precision so I'll be changing over to `integer`.

Comment (spikecurtis) [reply coderd/database/migrations/000223_notifications_metrics.up.sql]: This is backing a Prometheus metric, which are always floats, and by convention end in the SI unit, which for time is seconds.  It implies no precision.

This should be a float: there just isn't any compelling reason to pre-judge the required precision. This is not a banking application, nor do we need to worry about the relative speed of integer vs floating point arithmetic.

The name `queued_seconds` is appropriate.

File: coderd/notifications/metrics.go
Comment (spikecurtis) [inline coderd/notifications/metrics.go:42]: DispatchCount, TempFailureCount, and PermFailureCount might make sense as a single metric with different "disposition" (or "result") labels.  The help text refers to them as "notifications", but I think that's misleading.  What we are counting is "notification attempts" or "dispatch attempts", which might individually end up in success or some kind of failure.  In particular, in the case of failure and retry, the number of delivery attempts can exceed the number of notifications, so it's important to make the distinction.

Prometheus guidelines say that something should be a single metric with labels (vs multiple metrics) when the sum or average of the label is sensible and coherent. Filtering by method and summing across disposition would be useful, for example, to track the number of web requests we are making via Webhook delivery.

File: coderd/notifications/metrics.go
Comment (spikecurtis) [inline coderd/notifications/metrics.go:69]: A really useful Gauge is the number of "in progress" delivery attempts.  You increment it when you kick off the delivery goroutine and decrement it when you send to `success` or `failure` channels.  Can help spot hung handlers/leaked goroutines

File: coderd/notifications/metrics.go
Comment (spikecurtis) [inline coderd/notifications/metrics.go:32]: accumulating counts should end in the units and _total_ as in `coderd_notifications_successful_deliveries_total`, where "deliveries" is the unit (could use "dispatch_attempts", or something similar).

https://prometheus.io/docs/practices/naming/

File: coderd/notifications/metrics.go
Comment (spikecurtis) [inline coderd/notifications/metrics.go:68]: Sounds too generic to me.  "The number of delivery attempt results waiting to be flushed to the store" ?

File: coderd/database/migrations/000223_notifications_metrics.up.sql
Comment (spikecurtis) [inline coderd/database/migrations/000223_notifications_metrics.up.sql:2]: Do we really to store this on the row?  It feels a little strange and ambiguous.  If the notification_message is currently queued, then it's a bit meaningless, since we don't know for how much longer it will be queued. I.e. it is at present being returned on Enqueue, but is meaningless in that context.

When we dequeue it, can't the notifier compute how long it was on the queue based on it's `created_at` or `updated_at` time?  Or, if we really want to ensure all the measurements are based on the central PostgreSQL server clock, then it can part of the result returned to the notifier without having to write it back to the database.

File: coderd/notifications/notifier.go
Comment (spikecurtis) [inline coderd/notifications/notifier.go:246]: I think it would be cleaner to just compute this metric on-`Collect`, rather than `Set`ing it when something reads or writes to these channels.  As an illustration of how easy it is to miss a spot, note that you have forgotten to set it in `process()` after a failed call to `prepare()`.

File: coderd/notifications/metrics.go
Comment (spikecurtis) [inline coderd/notifications/metrics.go:51]: We say this should be less than the fetch interval, which is 15s --- so I think there should be more buckets <15s.  In fact, it should be more like 15/(n+1) where n is the number of Coderd replicas, assuming that notifications usually get picked up by the next fetch.  Seeing the average move up toward 15s is way more important than the distinction between 10minutes, 1 hour and 1 day, all of which indicate that notifications are just blocked entirely.

File: coderd/notifications/metrics.go
Comment (spikecurtis) [inline coderd/notifications/metrics.go:61]: should be `0.001, 0.005, 0.01, 0.05` on the bottom end, right?

File: coderd/notifications/metrics.go
Comment (spikecurtis) [inline coderd/notifications/metrics.go:69]: Also a count of success/failure updates flushed to the database.

File: coderd/notifications/notifier.go
Comment (dannykopping) [inline coderd/notifications/notifier.go:237]: Addresses https://github.com/coder/coder/pull/13537#discussion_r1668844087

File: coderd/database/queries/notifications.sql
Comment (mafredri) [inline coderd/database/queries/notifications.sql:22]: I think you can revert the hack if you change this to `RETURNING id::uuid;`. Alternatively, in `VALUES(`, using `@id::uuid` could also work.

File: coderd/notifications/metrics_test.go
Comment (mafredri) [inline coderd/notifications/metrics_test.go:102]: Linter unhappy, this should be fine with Go 1.22, though.

**Review by mtojek — APPROVED** [review]
👍

File: coderd/database/queries/notifications.sql
Comment (mafredri) [reply coderd/database/queries/notifications.sql:22]: Actually, I may be wrong. Not really sure what you ran into 😄

**Review by mafredri — APPROVED** [review]

File: coderd/database/queries/notifications.sql
Comment (dannykopping) [reply coderd/database/queries/notifications.sql:22]: Oh nice! That worked 🥇

File: coderd/coderd.go
Comment (dannykopping) [inline coderd/coderd.go]: Whoops!

**Review by mtojek — APPROVED** [review]
Stamp 👍

File: coderd/notifications/metrics.go
Comment (spikecurtis) [reply coderd/notifications/metrics.go:66]: Dispatching, say, a webhook, within the same cluster could be a few milliseconds latency.  Same geographic region could be 10-20ms.

Almost nothing is going to be 1ms unless it's extremely simple and extremely close, but in 50ms you could be dispatching across an entire continent.  With buckets like this you could have something that normally returns in 15ms take 3x its normal time and be close to falling over and this metric would not clock it.  That seems much more important than 30s, 60s, 120s.

Comment (dannykopping) [reply coderd/notifications/metrics.go:66]: I hear you and I can see where you're coming from.

Let's wait for users to ask for this. I think there's sufficient coverage for now, and the additional datapoints are likely not going to be very useful for the majority of operators IMHO.

**Pull request merged**
